scio-core/src/main/scala/com/spotify/scio/io/ZstdDictIO.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.io
+
+import com.github.luben.zstd.ZstdDictTrainer
+import com.spotify.scio.ScioContext
+import com.spotify.scio.coders.CoderMaterializer
+import com.spotify.scio.values.SCollection
+import org.apache.beam.sdk.util.CoderUtils
+import org.slf4j.LoggerFactory
+
+import scala.util.{Random, Try}
+
+case class ZstdDictIO[T](path: String) extends ScioIO[T] {
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  override type ReadP = Nothing // WriteOnly
+  override type WriteP = ZstdDictIO.WriteParam
+
+  override val tapT: TapT.Aux[T, Nothing] = EmptyTapOf[T]
+
+  override protected def read(sc: ScioContext, params: Nothing): SCollection[T] =
+    throw new UnsupportedOperationException("ZstdDictIO is write-only")
+
+  override protected def write(
+    data: SCollection[T],
+    params: ZstdDictIO.WriteParam
+  ): Tap[Nothing] = {
+    val ZstdDictIO.WriteParam(
+      zstdDictSizeBytes,
+      numElementsForSizeEstimation,
+      trainingBytesTarget
+    ) = params
+    // see https://github.com/facebook/zstd/issues/3769#issuecomment-1730261489
+    if (zstdDictSizeBytes > (ZstdDictIO.DictionarySizeMbWarningThreshold * 1024 * 1024)) {
+      logger.warn(
+        s"Dictionary sizes over ${ZstdDictIO.DictionarySizeMbWarningThreshold}MB are of " +
+          s"questionable utility. Consider reducing zstdDictSizeBytes."
+      )
+    }
+    if (numElementsForSizeEstimation <= 0) {
+      throw new IllegalArgumentException(
+        s"numElementsForSizeEstimation must be positive, found $numElementsForSizeEstimation"
+      )
+    }
+    // training bytes may not exceed 2GiB a.k.a. the max value of an Int
+    val trainingBytesTargetActual: Int = Option(trainingBytesTarget).getOrElse {
+      val computed =
+        Try(Math.multiplyExact(zstdDictSizeBytes, 100)).toOption.getOrElse(Int.MaxValue)
```

Comment:
do we want to catch this or let it fail ?

Reply:
hmm actually I am unsure of what happens in this case, like if the training size is related to the elements added or the dict size itself. 2gb / 100 implies that the max zdict dictionary size should be 20ish megs, so maybe we should throw on that

Reply:
+1 on throwing an exception

Reply:
Re-reading, ~100x is just a recommendation, so using fewer items would probably decrease its effectiveness but wouldn't be catastrophic. Gonna throw anyway since recommendations argue against large dictionaries.